inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/ClassRetransformationService.java

```diff
@@ -0,0 +1,145 @@
+ * Before a class is submitted for retransformation, it is again checked if the retransformation is actually required.
+ */
+@Service
+@Slf4j
+public class ClassRetransformationService {
+
+    @Autowired
+    private BatchJobExecutorService executor;
+
+    @Autowired
+    private InspectitEnvironment env;
+
+    @Autowired
+    private List<SpecialSensor> specialSensors;
+
+    @Autowired
+    private SelfMonitoringService selfMonitoring;
+
+    @Autowired
+    private Instrumentation instrumentation;
+
+    @Autowired
+    private AsyncClassTransformer transformer;
+
+    /**
+     * The queue of classes to process.
+     * For each element in the queue, this service checks if a retransformation is still required.
+     * If this is the case, then a retransformation is triggered.
+     */
+    private Set<Class<?>> pendingClasses = Collections.synchronizedSet(Collections.newSetFromMap(new WeakHashMap<>()));
+
+    /**
+     * The background job repeatedly executing {@link #retransformClasses(int)}
+     */
+    private BatchJobExecutorService.BatchJob retransformJob;
+
+    /**
+     * Calling this mehtod lets this service know that the instrumentation configuration for the given class has changed and that it requires a retransformation.
+     *
+     * @param clazz the clazz to put to the working queue
+     */
+    public void requestRetransformation(Class<?> clazz) {
+        pendingClasses.add(clazz);
+    }
+
+    @PostConstruct
+    private void init() {
+        val settings = env.getCurrentConfig().getInstrumentation().getInternal();
+        Duration delay = settings.getClassConfigurationCheckInterBatchDelay();
+        int batchSize = settings.getClassRetransformBatchSize();
+        retransformJob = executor.startJob(this::retransformClasses, batchSize, delay, delay);
+    }
+
+    /**
+     * Reconfigures the jobs batch sizes and the inter batch delay on configuration changes.
+     *
+     * @param ev the change event
+     */
+    @EventListener
+    private void configEventListener(InspectitConfigChangedEvent ev) {
+        val settings = ev.getNewConfig().getInstrumentation().getInternal();
+        retransformJob.setBatchSizes(settings.getClassRetransformBatchSize());
+        retransformJob.setInterBatchDelay(settings.getClassRetransformInterBatchDelay());
+    }
+
+    /**
+     * This method is regularly called through {@link #retransformJob} to process elements from {@link #pendingClasses}.
+     * The goal is to get and remove batchSize classes from the queueand to invoke {@link Instrumentation#retransformClasses(Class[])} on them.
+     * <p>
+     * The package access is for testing.
+     */
+    void retransformClasses(int batchSize) {
+        val watch = new StopWatch();
+        if (!transformer.isShuttingDown()) {
+            Set<Class<?>> classesToRetransform = new HashSet<>();
+            try {
+                Optional<Class<?>> nextClass = CommonUtils.pollElementFromSet(pendingClasses);
```

Comment:
does it make sense to pull one by one, how about we pull at least the `bacthSize`?

Reply:
Set does not offer this functionality, however with fuava sets we can reuse the iterator.